lib/SIL/Utils/BasicBlockUtils.cpp

```diff
@@ -372,3 +372,146 @@ void DeadEndBlocks::compute() {
       ReachableBlocks.insert(Pred);
   }
 }
+
+//===----------------------------------------------------------------------===//
+//                    Post Dominance Computation Utilities
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+struct PostDominanceCompletionSetComputer {
+  SILBasicBlock *definingBlock;
+  SILBasicBlock *reachableBlock;
+  SmallPtrSetImpl<SILBasicBlock *> &visitedBlocks;
+  DeadEndBlocks &deBlocks;
+
+  SmallVector<SILBasicBlock *, 32> worklist;
+  SmallSetVector<SILBasicBlock *, 8> successorBlocksThatMustBeVisited;
+
+  /// Set to true if we found a loop while processing.
+  bool foundLoop = false;
+
+  PostDominanceCompletionSetComputer(
+      SILBasicBlock *definingBlock, SILBasicBlock *reachableBlock,
+      SmallPtrSetImpl<SILBasicBlock *> &visitedBlocks,
+      DeadEndBlocks &deadEndBlocks)
+      : definingBlock(definingBlock), reachableBlock(reachableBlock),
+        visitedBlocks(visitedBlocks), deBlocks(deadEndBlocks) {
+    assert(definingBlock != reachableBlock);
+  }
+
+  void performDataflow();
+  void initializeDataflow();
+};
+
+} // anonymous namespace
+
+void PostDominanceCompletionSetComputer::initializeDataflow() {
+  visitedBlocks.insert(reachableBlock);
+  for (auto *pred : reachableBlock->getPredecessorBlocks()) {
+    // Check if we have an over consume.
+    if (pred == reachableBlock) {
+      for (auto *succ : pred->getSuccessorBlocks()) {
+        if (!visitedBlocks.count(succ))
+          successorBlocksThatMustBeVisited.insert(succ);
+      }
+      foundLoop = true;
+      continue;
+    }
+
+    worklist.push_back(pred);
+  }
+}
+
+void PostDominanceCompletionSetComputer::performDataflow() {
```

Comment:
I guess you could call the "data flow" but it's a little confusing because all it's doing is marking blocks visited. It just finds reachable blocks.